sendWithAsyncReply reply lambdas should have optional Connection parameter

achristensen07 opened the pull request:
#### 88b877c5ccdb356ac1be2b3614c7f167eeb14a74
<pre>
sendWithAsyncReply reply lambdas should have optional Connection parameter
<a href="https://bugs.webkit.org/show_bug.cgi?id=290144">https://bugs.webkit.org/show_bug.cgi?id=290144</a>
<a href="https://rdar.apple.com/147543627">rdar://147543627</a>

Reviewed by Chris Dumez.

When we receive IPC, we receive it from a Connection.  We already have logic in HandleMessage.h
that uses sfinae to add the Connection as the first parameter if the receiver has a Connection*
parameter, like WebPageProxy::pageScaleFactorDidChange.

This PR makes a similar mechanism for the reply of sendWithAsyncReply.  We now have the ability
to just put a Connection* parameter in the reply and it&apos;ll just be called with the connection
that sent the reply.  This allows us to message check the correct connection, for example.

There is an edge case when AuxiliaryProcessProxy::sendWithAsyncReply is called and the process
has either already terminated or it is waiting for its process to start and it never successfully
starts.  In this case, there is no Connection to cancel the reply with.  Because this case is
unavoidable, I made it so we always call the completion handler with a null connection when we
are cancelling a reply.  At that point, even if we do have a connection, the connection is on
its way to its end anyways, so there is no point in message checking it and there&apos;s nothing else
we can do with the connection.  I made the message checks more tolerant of null connections too.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::createEmptyForAuxiliaryProcessReply):
(IPC::Connection::Connection):
(IPC::Connection::invalidate):
(IPC::Connection::sendMessageImpl):
(IPC::Connection::sendMessageWithAsyncReply):
(IPC::Connection::sendMessageWithAsyncReplyWithDispatcher):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::dispatchIncomingMessageForTesting):
(IPC::Connection::dispatchOnReceiveQueueForTesting):
(IPC::Connection::dispatchMessage):
(IPC::Connection::cancelAsyncReplyHandlers):
(IPC::CompletionHandler&lt;void):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
(IPC::CompletionHandler&lt;void):
(IPC::Connection::makeAsyncReplyHandlerWithDispatcher):
(IPC::callWithConnectionAndArgsTuple):
(IPC::Connection::callReply):
(IPC::Connection::cancelReply):
(IPC::Connection::protectedConnectionQueue const): Deleted.
* Source/WebKit/Platform/IPC/MessageSenderInlines.h:
(IPC::cancelReplyWithoutUsingConnection):
(IPC::callReplyWithoutUsingConnection):
(IPC::MessageSender::sendWithAsyncReplyWithoutUsingIPCConnection const):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::sendWithAsyncReply):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformOpen):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::sendMessage):
(WebKit::AuxiliaryProcessProxy::replyToPendingMessages):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getNetworkProcessConnection):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendGestureEvent):
(WebKit::WebPageProxy::sendPreventableTouchEvent):
(WebKit::WebPageProxy::handlePreventableTouchEvent):
(WebKit::WebPageProxy::handleTouchEvent):
(WebKit::WebPageProxy::didReceiveEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.mm:
(WebKit::TiledCoreAnimationDrawingAreaProxy::dispatchPresentationCallbacksAfterFlushingLayers):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsSendWithAsyncReply):
(WebKit::IPCTestingAPI::jsWaitForAsyncReplyAndDispatchImmediately):

Canonical link: <a href="https://commits.webkit.org/292508@main">https://commits.webkit.org/292508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13d2e23bf0fb87a025f393fa187d67fc6570d8e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101295 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46747 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16138 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24275 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73359 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30587 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12112 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86937 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53698 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11863 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81984 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103322 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23295 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16965 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82399 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23546 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82956 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81775 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26397 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3840 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16669 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15490 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23258 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22917 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26397 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24658 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
Personally, I would have preferred to pass a null Connection rather than an empty one. Using a raw pointer makes it clear that it can be null (in case of failure) and it lets the client properly deal with this case. Now they're getting a Connection object that does nothing and it is not directly obvious how to know if the connection is empty/valid or not.

I do not feel super strongly if you want to land this way, just my opinion.

achristensen07 committed:
sendWithAsyncReply reply lambdas should have optional Connection parameter
https://bugs.webkit.org/show_bug.cgi?id=290144
rdar://147543627

Reviewed by Chris Dumez.

When we receive IPC, we receive it from a Connection.  We already have logic in HandleMessage.h
that uses sfinae to add the Connection as the first parameter if the receiver has a Connection*
parameter, like WebPageProxy::pageScaleFactorDidChange.

This PR makes a similar mechanism for the reply of sendWithAsyncReply.  We now have the ability
to just put a Connection* parameter in the reply and it'll just be called with the connection
that sent the reply.  This allows us to message check the correct connection, for example.

There is an edge case when AuxiliaryProcessProxy::sendWithAsyncReply is called and the process
has either already terminated or it is waiting for its process to start and it never successfully
starts.  In this case, there is no Connection to cancel the reply with.  Because this case is
unavoidable, I made it so we always call the completion handler with a null connection when we
are cancelling a reply.  At that point, even if we do have a connection, the connection is on
its way to its end anyways, so there is no point in message checking it and there's nothing else
we can do with the connection.  I made the message checks more tolerant of null connections too.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::createEmptyForAuxiliaryProcessReply):
(IPC::Connection::Connection):
(IPC::Connection::invalidate):
(IPC::Connection::sendMessageImpl):
(IPC::Connection::sendMessageWithAsyncReply):
(IPC::Connection::sendMessageWithAsyncReplyWithDispatcher):
(IPC::Connection::processIncomingMessage):
(IPC::Connection::dispatchIncomingMessageForTesting):
(IPC::Connection::dispatchOnReceiveQueueForTesting):
(IPC::Connection::dispatchMessage):
(IPC::Connection::cancelAsyncReplyHandlers):
(IPC::CompletionHandler<void):
* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::waitForAsyncReplyAndDispatchImmediately):
(IPC::CompletionHandler<void):
(IPC::Connection::makeAsyncReplyHandlerWithDispatcher):
(IPC::callWithConnectionAndArgsTuple):
(IPC::Connection::callReply):
(IPC::Connection::cancelReply):
(IPC::Connection::protectedConnectionQueue const): Deleted.
* Source/WebKit/Platform/IPC/MessageSenderInlines.h:
(IPC::cancelReplyWithoutUsingConnection):
(IPC::callReplyWithoutUsingConnection):
(IPC::MessageSender::sendWithAsyncReplyWithoutUsingIPCConnection const):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::sendWithAsyncReply):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformOpen):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::sendMessage):
(WebKit::AuxiliaryProcessProxy::replyToPendingMessages):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getNetworkProcessConnection):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendGestureEvent):
(WebKit::WebPageProxy::sendPreventableTouchEvent):
(WebKit::WebPageProxy::handlePreventableTouchEvent):
(WebKit::WebPageProxy::handleTouchEvent):
(WebKit::WebPageProxy::didReceiveEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.mm:
(WebKit::TiledCoreAnimationDrawingAreaProxy::dispatchPresentationCallbacksAfterFlushingLayers):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsSendWithAsyncReply):
(WebKit::IPCTestingAPI::jsWaitForAsyncReplyAndDispatchImmediately):

Canonical link: https://commits.webkit.org/292508@main
